Sources/LiveKit/Core/SignalClient.swift

```diff
@@ -608,6 +617,70 @@ internal extension SignalClient {
 
         return sendRequest(r)
     }
+
+    @discardableResult
+    private func sendPing() -> Promise<Void> {
+        log("ping/pong: sending...")
+
+        let r = Livekit_SignalRequest.with {
+            $0.ping = Int64(Date().timeIntervalSince1970)
+        }
+
+        return sendRequest(r)
+    }
+}
+
+// MARK: - Server ping/pong logic
+
+private extension SignalClient {
+
+    func onPingIntervalTimer() {
+
+        guard let jr = latestJoinResponse else { return }
+
+        pingIntervalTimer?.suspend()
+        sendPing()
+
+        pingTimeoutTimer = {
+            let timer = DispatchQueueTimer(timeInterval: TimeInterval(jr.pingTimeout), queue: queue)
+            timer.handler = { [weak self] in
+                guard let self = self else { return }
+                self.log("ping/pong timed out", .error)
+                self.cleanUp(reason: .networkError(SignalClientError.serverPingTimedOut()))
+            }
+            timer.resume()
+            return timer
+        }()
+    }
+
+    func onReceivedPong(_ r: Int64) {
+
+        log("ping/pong received pong from server")
+        // clear timeout timer
+        pingTimeoutTimer = nil
+        // carry on...
+        pingIntervalTimer?.resume()
+    }
+
+    func restartPingTimer() {
+        // always suspend first
+        pingIntervalTimer = nil
+        pingTimeoutTimer = nil
+        // check received joinResponse already
+        guard let jr = latestJoinResponse,
+              // check server supports ping/pong
+              jr.pingTimeout > 0,
+              jr.pingInterval > 0 else { return }
+
+        log("ping/pong starting with interval: \(jr.pingInterval)")
+
+        pingIntervalTimer = {
```

Comment:
I don't think it would be a big problem, but just to make sure I understand correctly, the flow with this implementation is. 
1. Join
2. restartPingTimer 
3. send ping for the first time after pingInterval time elapsed
4. only start pingTimeout if a pong has been received

so there's no actual interval for pings that keeps running, but rather every single ping message relies on a pong message to be received, right?

Reply:
Yes the ping will not be sent out until pong arrives. `pingTimeoutTimer` will close the connection if pong doesn't arrive.

I could change the logic to continuously send out pings.